### DISCUSSION TITLE
Pairing problem - home app

### DISCUSSION THREAD

**Discussion by aLAN-LDZ** [HomeSpan/HomeSpan#15] [Q&A] [answered]
Hi

At first I wanted to say - great library!

I was able to connect to wifi, I can see the accessory in the home application, unfortunately, after entering the code, it shows "cannot connect to the accessory"

I am using the esp32-wroom-32d module:

https://www.tme.eu/pl/details/esp32-devkitc-32d/zestawy-do-transmisji-danych/espressif/?brutto=1&gclid=CjwKCAiAudD_BRBXEiwAudakXynY4jtcNNcGGtab2MRXTiuwui7MfCaxY-22NvjhuAtPoBymP9v-bRoCWZ8QAvD_BwE

**Comment by HomeSpan** [comment]
What do the log files show?  You can turn on log details from the Arduino Serial Monitor by typing 'L2 <return>' (which sets the log level to 2, the maximum).  In particular, we want to check that HomeSpan is indeed receiving the request for pairing.

**Comment by aLAN-LDZ** [comment]
Thank you for your answer. This is what it looks like in the logs:

`Connecting... Success!
IP: 10.9.137.86

Starting MDNS...
Broadcasting as: HomeSpan-820F1E5E7975.local (HomeSpan Ceiling Fan / HomeSpan-ESP32)

Starting Web (HTTP) Server supporting up to 8 simultaneous connections...

DEVICE NOT YET PAIRED -- PLEASE PAIR WITH HOMEKIT APP

HomeSpan Ceiling Fan is READY!


*** Log Level set to 2

=======================================
** Client #0 Connected: (14 sec) 10.5.71.171

<<<<<<<<< 10.5.71.171 <<<<<<<<<
POST /pair-setup HTTP/1.1
Host: HomeSpan\032Ceiling\032Fan._hap._tcp.local
Content-Length: 6
Content-Type: application/pairing+tlv8
------------ END BODY! ------------
STATE(1) 01
METHOD(1) 00
------------ END TLVS! ------------
In Pair Setup...Found <M1>.  Expected <M1>

>>>>>>>>>> 10.5.71.171 >>>>>>>>>>
HTTP/1.1 200 OK
Content-Type: application/pairing+tlv8
Content-Length: 409

STATE(1) 02
PUBKEY(384) 9B7FA6CB32DDC97404502F9D0D81543E15F5999885CAAEA853F1632A63A327DE240D2539B37AE12AF35EA1F9F9C7C455EEC16A6DE10CEE5C4E8DC7C72BF0BC8682F503A9603B3491FFD7B954F78ADFA95747A3AD750CD8DF6BBB85A535BE4BBEE48A32B66C86C926A1213C238B72CFA7A55041E15E82D7AB231A1A016E0A456FEEEE15F90D53106F5D3587A0B458ABE6CB3670B25E676A3F54DBB34654B8DAC3FE4D2B068E6712058E112C6E36DFC5AB7EEC890ECA6606BFC9E9C4EB834F863D587CA21456A7C949488AE5B1B0BCEFE914BFF31F5A22D32CA8ACEA634AA6BD00B2976A72E66F31CFF33CAAC21123B3F5A59499DBBB50D6E16B497D812308DF7D4B50BC340874E2E46D2BE47D122CE2B3FDF14C566B5354F4441E2A681B74C31C180E2526CA8E997F870564DB239685EF0A436276794EFB2B857F8E06C549571E6D6894C47BD15FE0F4822D40DCE60C7683E3E6A2F5EF9C7FE891B4F5614BAB05C34615F9F6FF99ED24A9D66C14565CBE5DBAD7D9EB5F6A58F969789A22149932
SALT(16) CF961A3DC675EE63DB33650BDC9D75D6
------------ SENT! --------------
** Disconnecting Client #0  (16 sec)`

I also noticed that the accessory is detected only for a few moments after the restart. then before I enter the code on the phone - it's already disconnected

**Comment by HomeSpan** [comment]
Hmm.  Everything looks correct except for the very last line that says "Disconnecting."  This is reported only if the ESP32 is dropping the connection, but the code keeps the connection open in this part of the sequence.  However, it does check to see that the connection is still live, which means that for some reason the ESP32 is dropping the connection after the sending the packets to the phone.  This is consistent with what you are seeing on the phone - it receives the packets and the ESP32 should be standing by to get the code you type into the phone.  But instead of standing by, it's dropping the connection.

It may be that this is something to do with the WROOM-32D (which has been deprecated by Espressif) and the way it handles WiFi?  (I'm running on the WROOM-32, which is slightly different).  The only other thing we can check is the version of the ESP32 IDF used to compile the software.  When you first start the ESP32, the initial messages should show the ESP-IDF Version.  I'm running on v3.2.3-14.  Also, are you compiling under the Arduino IDE?

**Reply by aLAN-LDZ** [reply]
I don't know how to check ESP-IDF but it compiles with the newest ArduinoIDE :(

**Comment by HomeSpan** [comment]
The ESP-IDF is listed at the top of the log files when you first run HomeSpan.  If you reset the ESP32 while the Serial Monitor is open you should see it listed as part of the initial output, along with a number of other status and welcome messages.

**Reply by aLAN-LDZ** [reply]
Thanks the wersion of ESP-IDF  v3.2.3-14-gd3e562907

**Comment by HomeSpan** [comment]
Okay, that matches my version so is not the source of the problem.  On the off-chance there is something the iPhone does not like about the specific Public Key the ESP32 is generating, we can have it generate a new one and a new ID.  To do so, type 'H <return>' from within the Serial Monitor.  This will delete the internal HomeKit ID and restart the ESP32, at which point a new ID is randomly generated. This will not change any of the WiFi Credentials.   Then try pairing again.  I doubt this will work, but its work a try. 

Also, are you running the latest versions of iOS on your iPhone?

**Reply by aLAN-LDZ** [reply]
Thank you for your commitment and willingness to help. Unfortunately, pressing H did not help :(.

Yes I have the latest iOS, I was looking at the iPad too.

I noticed that the accessory is only detected a few seconds after rebooting esp later it is no longer visible

**Comment by HomeSpan** [comment, marked as answer]
That's puzzling, and suggests that the WiFi connection itself may be dropping, or that there may be a problem with the MDNS protocol that broadcasts the presence of the ESP32 and HomeSpan.  Here is another thing you can try to see if the MDNS is a problem:  There is an App on the iPhone App Store called *Discovery-DNS-SD Browser*.  If you search for the words "discovery" and "bonjour" together, it should show up.  This app allows you to read all bonjour devices.  Safari previously had this ability but it was deleted for some reason a few years ago.

Once you install the App, run it and select "local" as the domain (it will probably be the only selection).  This will bring up a long list of bonjour MDNS services.  Scroll down until you find something labeled "_hap._tcp."  There will also be some selections that say "homekit" that are not relevant.  Once you select "_hap._tcp" you should see all device on your network that are broadcasting HomeKit services.  One of the services should say HomeSpan (or HomeSpan Ceiling Fan).   If you don't see this, that means the ESP32 is not broadcasting MDNS.  Or, if you see it when you first boot the ESP32, but it disappear in a bit, the only thing I can think of is that maybe your WiFi router is filtering something?  The ESP32 should be broadcasting MDNS as soon as its connected to WiFi, and during all operations.

**Reply by aLAN-LDZ** [reply]
Like you say, I only see him for a moment. Unfortunately, I cannot access the router. As soon as I have another one on hand, I will let you know if this has solved the problem. Thank you for your help

**Reply by ZhouMinChuan** [reply]
I also have this problem! The "_hap._tcp" just appears for a while.  So how can I set up my wifi router to deal with this problem? Thank you

**Reply by HomeSpan** [reply]
What version of the Arduino-ESP32 board manager are you using?  Espressif made changes to the MDNS code starting in version 2.0.1 that seems to have addressed the issue for a number of users.

**Reply by ZhouMinChuan** [reply]
I update my Arduino-ESP32 board manager to version 2.0.3 but I still can't connect to my iPhone. After scanning the Setup Code also _unable to add accessory_

**Reply by HomeSpan** [reply]
Are you testing with basic, unmodified example, such as Tutorial Example 1.  See #338 for more details.

**Reply by vgataullin** [reply]
I have found a solution for my Asus router: https://superuser.com/questions/1238812/how-to-check-if-a-router-support-bonjour-protocol

On ASUSWRT based Asus routers;

To disable "AirTime Fairness", you can;

Click “Wireless” on left
Click “Professional” tab
Set “Band” to “2.4GHz”
Set “Airtime Fairness” to “Disable”
Click “Apply”
NB: If your printer can also connect to WiFi using 5GHz, you'll need to repeat the steps above, while setting the band to 5GHz instead.

**Comment by aLAN-LDZ** [comment]
You were right :) it works fine on a normal router. This Tobic can be closed. I have another problem, but that's on a separate topic.

**Comment by xandermeyers** [comment]
I have an esp32 and I cant Connect with my iPhone on de home app. 1) esp32 can Connect with the wifi 2) wifi is stable and strong 3) esp32 is visable on the home app 

Once i want to Connect with the esp32, it will load for a min or 2 and after that I got the message cant Connect with Accessory.

I have a code one code that works fine and one that has this problem. The only difference in the code is the way how I let the esp32 Connect with the wifi.

Code that works: I have to put the wifi and password in the seriële monitor

code that doesnt work: esp32 sends out a wifi —> my phone Connects to that wifi —> via my phone I put the wifi and the password —> esp32 connected to the wifi.

both codes have no problem with the wifi, only to Connect with the home app.

Hopefully somebody can help me😁

1) doesnt work
#include <WiFi.h>
#include <WebServer.h>
#include "HomeSpan.h"
#include "LED.h"
#include "LED1.h"
#include "LED2.h"
#include "LED3.h"
#include "LED4.h"
#include "LED5.h"
#include "LED6.h"
#include "LED7.h"
#include "LED8.h"
#include "LED9.h"


const char *softAP_SSID = "ESP32-Setup";
const char *softAP_Password = "12345678";
WebServer server(80);

bool connectedToWiFi = false;

const char *htmlForm = "<html><body><h1>Wi-Fi Configuratie</h1>"
                       "<form action='/connect' method='POST'>"
                       "<label for='ssid'>SSID:</label><input type='text' id='ssid' name='ssid'><br>"
                       "<label for='password'>Wachtwoord:</label><input type='password' id='password' name='password'><br>"
                       "<input type='submit' value='Verbinding maken'>"
                       "</form></body></html>";

void handleRoot() {
  server.send(200, "text/html", htmlForm);
}

void handleConnect() {
  String ssid = server.arg("ssid");
  String password = server.arg("password");

  Serial.print("SSID ontvangen: ");
  Serial.println(ssid);
  Serial.print("Wachtwoord ontvangen: ");
  Serial.println(password);

  WiFi.begin(ssid.c_str(), password.c_str());
  server.send(200, "text/html", "<html><body><h1>Verbinden...</h1><p>Even geduld aub.</p></body></html>");

  int tries = 0;
  while (WiFi.status() != WL_CONNECTED && tries < 40) {
    delay(500);
    Serial.print(".");
    tries++;
  }

  if (WiFi.status() == WL_CONNECTED) {
    Serial.println("\n✅ Verbonden!");
    Serial.print("IP-adres: ");
    Serial.println(WiFi.localIP());
    connectedToWiFi = true;

    WiFi.softAPdisconnect(true);
    delay(1000);

    // ⬇️ Wacht even voordat HomeSpan start (optie 3)
    delay(2000);

    setupHomeSpan();

  } else {
    Serial.println("\n❌ Verbinden mislukt. Herstart ESP in 5 seconden...");
    delay(5000);
    ESP.restart();
  }
}

void setupHomeSpan() {
  Serial.println("HomeSpan begint...");
  homeSpan.setPairingCode("11223344"); // Je pairingcode hier
  homeSpan.setLogLevel(3);  // 0=none, 1=errors, 2=info, 3=verbose

  homeSpan.begin(Category::Lighting, "My home LED");
   Serial.println("HomeSpan is gestart, wacht op verbinding...");

  // RGB
  new SpanAccessory(1);  
    new Service::AccessoryInformation();
      new Characteristic::Identify();
      new Characteristic::Name("RGB LED");  
    new DEV_RgbLED(4, 2, 15); 
/*
  // LED1 t.e.m. LED9
  new SpanAccessory(2);
    new Service::AccessoryInformation();
    new Characteristic::Identify();
    new Characteristic::Name("LED1");
    new LED1(23, 19);

  new SpanAccessory(3);
    new Service::AccessoryInformation();
    new Characteristic::Identify();
    new Characteristic::Name("LED2");
    new LED2(23, 18);

  new SpanAccessory(4);
    new Service::AccessoryInformation();
    new Characteristic::Identify();
    new Characteristic::Name("LED3");
    new LED3(23, 5);  

  new SpanAccessory(5);
    new Service::AccessoryInformation();
    new Characteristic::Identify();
    new Characteristic::Name("LED4");
    new LED4(22, 19);  

  new SpanAccessory(6);
    new Service::AccessoryInformation();
    new Characteristic::Identify();
    new Characteristic::Name("LED5");
    new LED5(22, 18);  

  new SpanAccessory(7);
    new Service::AccessoryInformation();
    new Characteristic::Identify();
    new Characteristic::Name("LED6");
    new LED6(22, 5);  

  new SpanAccessory(8);
    new Service::AccessoryInformation();
    new Characteristic::Identify();
    new Characteristic::Name("LED7");
    new LED7(21, 19);  

  new SpanAccessory(9);
    new Service::AccessoryInformation();
    new Characteristic::Identify();
    new Characteristic::Name("LED8");
    new LED8(21, 18);  

  new SpanAccessory(10);
    new Service::AccessoryInformation();
    new Characteristic::Identify();
    new Characteristic::Name("LED9");
    new LED9(21, 5);  */
}

void setup() {
  Serial.begin(115200);
  delay(1000);

  Serial.println("🔧 Start configuratie...");

  WiFi.softAP(softAP_SSID, softAP_Password);
  Serial.print("📶 SoftAP gestart op IP: ");
  Serial.println(WiFi.softAPIP());

  server.on("/", HTTP_GET, handleRoot);
  server.on("/connect", HTTP_POST, handleConnect);
  server.begin();
}

void loop() {
  if (!connectedToWiFi) {
    server.handleClient();
  } else {
    homeSpan.poll();
  }
}

2) does work
#include "HomeSpan.h"
#include "LED.h"
#include "warmLED.k.h"
#include "warmLED.b.h"
#include "neon.h"


void setup() {
  // put your setup code here, to run once:
  Serial.begin(115200);
  homeSpan.setPairingCode("11223344");
  homeSpan.begin(Category::Lighting,"My home LED");

   new SpanAccessory(1);  
    new Service::AccessoryInformation();
      new Characteristic::Identify();
      new Characteristic::Name("RGB LED");  
    new DEV_RgbLED(32, 22, 23);  

  // Warm LED K
  new SpanAccessory(2);
    new Service::AccessoryInformation();
      new Characteristic::Identify();
      new Characteristic::Name("Warm LED K");
    new warmLEDk(18, 19);

  // Neon
  new SpanAccessory(3);
    new Service::AccessoryInformation();
      new Characteristic::Identify();
      new Characteristic::Name("Neon");
    new neon(19);

  // Warm LED B
  new SpanAccessory(4);
    new Service::AccessoryInformation();
      new Characteristic::Identify();
      new Characteristic::Name("Warm LED B");
    new warmLEDb(5);
}

void loop() {
  // put your main code here, to run repeatedly:
 homeSpan.poll();
    
  }
